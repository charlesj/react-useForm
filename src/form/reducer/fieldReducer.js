import { List } from 'immutable'

export const actionTypes = {
  insertField: 'insertField',
  removeField: 'removeField',
  addListItem: 'addListItem',
  removeListItem: 'removeListItem',
  reset: 'reset',
  touched: 'touched',
  updateValue: 'updateValue',
  validationResult: 'validationResult',
}

export const actions = {
  insertField: (fieldName, fieldState) => ({ type: actionTypes.insertField, fieldName, payload: fieldState }),
  removeField: fieldName => ({ type: actionTypes.removeField, fieldName }),
  addListItem: (fieldName, fieldState) => ({ type: actionTypes.addListItem, fieldName, payload: fieldState }),
  removeListItem: (fieldName, index) => ({ type: actionTypes.removeListItem, fieldName, payload: index }),
  reset: (state) => ({ type: actionTypes.reset, payload: state }),
  touched: (fieldName) => ({ type: actionTypes.touched, fieldName }),
  updateValue: (fieldName, value) => ({ type: actionTypes.updateValue, fieldName, payload: value }),
  validationResult: (fieldName, error, helperText) => ({ type: actionTypes.validationResult, fieldName, payload: { error, helperText } }),
}

export const removedFieldsKey = 'removedFields'
export const fieldsKey = 'fields'

export const initState = state => {
  return state
}

const current = 'current'

export const fieldReducer = (state, { type, fieldName = '', payload }) => {
  let fieldPath = [fieldName]
  if (fieldName.includes('.items.')) {
    fieldPath = fieldName.split('.')
  }

  const handlers = {
    [actionTypes.updateValue]: value =>
      state.setIn([fieldsKey, ...fieldPath, current, 'value'], value)
        .setIn([fieldsKey, ...fieldPath, current, 'pristine'], value == state.getIn([fieldsKey, fieldName, 'initial', 'value'])),

    [actionTypes.touched]: () =>
      state.setIn([fieldsKey, ...fieldPath, current, 'touched'], true),

    [actionTypes.validationResult]: ({ error, helperText }) =>
      state.setIn([fieldsKey, ...fieldPath, current, 'error'], error)
        .setIn([fieldsKey, ...fieldPath, current, 'helperText'], helperText),

    [actionTypes.insertField]: fieldState =>
      state.setIn([fieldsKey, ...fieldPath], fieldState)
        .deleteIn([removedFieldsKey, state.get(removedFieldsKey, List()).indexOf(fieldName)]),

    [actionTypes.removeField]: () =>
      state.deleteIn([fieldsKey, ...fieldPath])
        .set(removedFieldsKey, state.get(removedFieldsKey, List()).push(fieldName)),

    [actionTypes.addListItem]: itemState => {
      const path = [fieldsKey, ...fieldPath, 'items']
      return state.setIn([...path, state.getIn(path, List()).size], itemState)
    },

    [actionTypes.removeListItem]: index =>
      state.deleteIn([fieldsKey, ...fieldPath, 'items', index]),

    [actionTypes.reset]: state => state,
  }

  const handler = handlers[type]
  if (handler) {
    return handler(payload)
  }
  return state
}
