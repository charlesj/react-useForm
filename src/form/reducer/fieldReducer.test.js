import { fromJS } from 'immutable'
import { fieldReducer, actions, removedFieldsKey, getFieldPath } from './fieldReducer'

let state, fieldName
beforeEach(() => {
  fieldName = 'field-name'
  state = fromJS({
    fields: {
      [fieldName]: {
        initial: {
          value: '',
        },
        current: {
          value: '',
          pristine: true,
          touched: false,
          error: false,
          helperText: '',
        },
      },
      otherField: {
        initial: {
          value: '',
        },
        current: {
          value: '',
          pristine: true,
          touched: false,
          error: false,
          helperText: '',
        },
      },
    },
  })
})

test('type not handled, return state', () => {
  expect(fieldReducer(state, {})).toEqual(state)
})

test('updateValue for field updates value on current', () => {
  const action = actions.updateValue(fieldName, 'new value')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('pristine is false if value not equal to initial', () => {
  const action = actions.updateValue(fieldName, 'new value')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('pristine is true if value equal to initial', () => {
  state = state.setIn([fieldName, 'current', 'pristine'], false)
  const action = actions.updateValue(fieldName, '')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('touched set on current field', () => {
  const action = actions.touched(fieldName)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('validationResult set on current field', () => {
  const action = actions.validationResult(fieldName, true, 'i am error')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('insertField', () => {
  const fieldState = fromJS({
    initial: {
      value: 'one',
    },
    current: {
      value: 'one',
    },
  })
  const action = actions.insertField('newField', fieldState)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('insertField removes fieldName from removedFields', () => {
  const fieldState = fromJS({
    initial: {
      value: 'one',
    },
    current: {
      value: 'one',
    },
  })
  const withRemoved = state.set(removedFieldsKey, fromJS(['newField']))
  const action = actions.insertField('newField', fieldState)
  expect(fieldReducer(withRemoved, action)).toMatchSnapshot()
})

test('removeField', () => {
  const action = actions.removeField(fieldName)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

describe('nested fields', () => {
  beforeEach(() => {
    state = fromJS({
      fields: {
        'parent.nested': {
          initial: {
            value: '',
          },
          current: {
            value: '',
            pristine: true,
            touched: false,
            error: false,
            helperText: '',
          },
        },
      },
    })
  })

  test('updateValue for field updates value on current', () => {
    const action = actions.updateValue('parent.nested', 'new value')
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

})

describe('list in state', () => {
  let listItemFieldName
  beforeEach(() => {
    listItemFieldName = `data.listField.items.1.fields.${fieldName}`
    state = fromJS({
      fields: {
        'data.listField': {
          items: [
            {},
            {
              fields: {
                [fieldName]: {
                  initial: {
                    value: '',
                  },
                  current: {
                    value: '',
                    pristine: true,
                    touched: false,
                    error: false,
                    helperText: '',
                  },
                },
              },
            },
          ],
        },
        otherField: {
          initial: {
            value: '',
          },
          current: {
            value: '',
            pristine: true,
            touched: false,
            error: false,
            helperText: '',
          },
        },
      },
    })
  })

  test('updateValue for field updates value on current', () => {
    const action = actions.updateValue(listItemFieldName, 'new value')
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  test('touched set on current field', () => {
    const action = actions.touched(listItemFieldName)
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  test('addListItem', () => {
    const fieldState = fromJS({
      fields: {
        [fieldName]: {
          initial: {
            type: 'text',
            value: '',
            optional: false,
            label: 'First',
            field: {
              label: 'First',
              name: 'data.listField.items.0.fields.fieldName',
            },
          },
          current: {
            helperText: '',
            error: false,
            pristine: true,
            touched: false,
            value: '',
          },
        },
      },
    })
    const action = actions.addListItem('data.listField', fieldState)
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  test('removeListItem', () => {
    const action = actions.removeListItem('data.listField', 1)
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

})

describe('getFieldPath', () => {
  test('single name', () => {
    expect(getFieldPath('name')).toEqual(['name'])
  })

  test('combined name', () => {
    expect(getFieldPath('data.name')).toEqual(['data.name'])
  })

  test('single name with items', () => {
    expect(getFieldPath('listField.items.1.fields.fieldName')).toMatchSnapshot()
  })

  test('combined name with items', () => {
    expect(getFieldPath('data.listField.items.1.fields.fieldName')).toMatchSnapshot()
  })
})
