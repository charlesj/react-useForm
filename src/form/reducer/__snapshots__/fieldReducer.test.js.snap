// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFieldPath combined name with items 1`] = `
Array [
  "data.listField",
  "items",
  "1",
  "fields",
  "fieldName",
]
`;

exports[`getFieldPath single name with items 1`] = `
Array [
  "listField",
  "items",
  "1",
  "fields",
  "fieldName",
]
`;

exports[`insertField 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "newField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "one",
      },
      "current": Immutable.Map {
        "value": "one",
      },
    },
  },
}
`;

exports[`insertField removes fieldName from removedFields 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "newField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "one",
      },
      "current": Immutable.Map {
        "value": "one",
      },
    },
  },
  "removedFields": Immutable.List [],
}
`;

exports[`list in state addListItem 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {},
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "value": "",
              },
              "current": Immutable.Map {
                "value": "",
                "pristine": true,
                "touched": false,
                "error": false,
                "helperText": "",
              },
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "type": "text",
                "value": "",
                "optional": false,
                "label": "First",
                "field": Immutable.Map {
                  "label": "First",
                  "name": "data.listField.items.0.fields.fieldName",
                },
              },
              "current": Immutable.Map {
                "helperText": "",
                "error": false,
                "pristine": true,
                "touched": false,
                "value": "",
              },
            },
          },
        },
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;

exports[`list in state removeListItem 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {},
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;

exports[`list in state touched set on current field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {},
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "value": "",
              },
              "current": Immutable.Map {
                "value": "",
                "pristine": true,
                "touched": true,
                "error": false,
                "helperText": "",
              },
            },
          },
        },
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;

exports[`list in state updateValue for field updates value on current 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {},
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "value": "",
              },
              "current": Immutable.Map {
                "value": "new value",
                "pristine": false,
                "touched": false,
                "error": false,
                "helperText": "",
              },
            },
          },
        },
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;

exports[`nested fields updateValue for field updates value on current 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "parent.nested": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "new value",
        "pristine": false,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;

exports[`pristine is false if value not equal to initial 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "new value",
        "pristine": false,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;

exports[`pristine is true if value equal to initial 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "field-name": Immutable.Map {
    "current": Immutable.Map {
      "pristine": false,
    },
  },
}
`;

exports[`removeField 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "removedFields": Immutable.List [
    "field-name",
  ],
}
`;

exports[`touched set on current field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": true,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;

exports[`updateValue for field updates value on current 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "new value",
        "pristine": false,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;

exports[`validationResult set on current field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": true,
        "helperText": "i am error",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
}
`;
